packages/workspace/src/parser/internal/native/lexer.ts

```diff
@@ -0,0 +1,170 @@
+  COMMENT: 'comment',
+  INVALID: 'invalid syntax',
+}
+
+export const rules: Record<string, moo.Rules> = {
+  main: {
+    [TOKEN_TYPES.MULTILINE_START]: { match: /'''[ \t]*[(\r\n)(\n)]/, lineBreaks: true, push: 'multilineString' },
+    [TOKEN_TYPES.DOUBLE_QUOTES]: { match: '"', push: 'string' },
+    [TOKEN_TYPES.NUMBER]: /-?(?:0|[1-9]\d*)(?:\.\d*)?(?:[eE][+-]?\d+)?/,
+    [TOKEN_TYPES.BOOLEAN]: /true|false/,
+    [TOKEN_TYPES.LEFT_PAREN]: '(',
+    [TOKEN_TYPES.RIGHT_PAREN]: ')',
+    [TOKEN_TYPES.ARR_OPEN]: '[',
+    [TOKEN_TYPES.ARR_CLOSE]: ']',
+    [TOKEN_TYPES.COMMA]: ',',
+    [TOKEN_TYPES.CCURLY]: '}',
+    [TOKEN_TYPES.OCURLY]: '{',
+    [TOKEN_TYPES.EQUAL]: '=',
+    [TOKEN_TYPES.WORD]: /[a-zA-Z_][\w.]*/s,
+    [TOKEN_TYPES.COMMENT]: /\/\//,
+    [TOKEN_TYPES.WHITESPACE]: { match: /[ \t]+/ },
+    [TOKEN_TYPES.NEWLINE]: { match: /[r\n]+/, lineBreaks: true },
+    [TOKEN_TYPES.INVALID]: { match: /[^ ]+/, error: true },
+  },
+  string: {
+    [TOKEN_TYPES.REFERENCE]: { match: /\$\{[ \t]*[\d\w.]+[ \t]*\}/, value: s => s.slice(2, -1).trim() },
+    [TOKEN_TYPES.DOUBLE_QUOTES]: { match: '"', pop: 1 },
+    [TOKEN_TYPES.CONTENT]: { match: /[^\\](?=")|[^\r\n]+?[^\\](?=\$\{|"|\n)/s, lineBreaks: false },
+    [TOKEN_TYPES.NEWLINE]: { match: /[r\n]+/, lineBreaks: true, pop: 1 },
+  },
+  multilineString: {
+    [TOKEN_TYPES.REFERENCE]: { match: /\$\{[ \t]*[\d\w.]+[ \t]*\}/, value: s => s.slice(2, -1).trim() },
+    [TOKEN_TYPES.MULTILINE_END]: { match: /^[ \t]*'''/, pop: 1 },
+    [TOKEN_TYPES.CONTENT]: { match: /.*\\\$\{.*[(\r\n)(\n)]|.*?(?=\$\{)|.*[(\r\n)(\n)]/, lineBreaks: true },
+  },
+}
+
+export class NoSuchElementError extends Error {
+  constructor(public lastValidToken?: LexerToken) {
+    super("All of the lexer's token has already been consumed.")
+  }
+}
+
+class InvalidLexerTokenError extends Error {
+  constructor() {
+    super('All lexer token must have a type.')
+  }
+}
+
+export type LexerToken = Required<moo.Token>
+
+const validateToken = (token?: moo.Token): token is LexerToken => {
+  if (token === undefined) {
+    return false
+  }
+  if (token.type === undefined) {
+    throw new InvalidLexerTokenError()
+  }
+  return true
+}
+
+class PeakableLexer {
+    private lexer: moo.Lexer
+    private peaked?: LexerToken
+    private peakedNoNewline?: LexerToken
+    constructor(src: string) {
+      this.lexer = moo.states(rules)
+      this.lexer.reset(src)
+    }
+
+    private advance(): LexerToken | undefined {
+      let token = this.lexer.next()
+
+      while (token && (token.type === TOKEN_TYPES.WHITESPACE
+        || token.type === TOKEN_TYPES.COMMENT)) {
+        if (token.type === TOKEN_TYPES.COMMENT) {
+          while (token && token.type !== TOKEN_TYPES.NEWLINE) {
+            token = this.lexer.next()
+          }
+        }
+        token = this.lexer.next()
+      }
+      return validateToken(token) ? token : undefined
+    }
+
+    public peak(ignoreNewlines = true): LexerToken | undefined {
```

Comment:
default access is public
```suggestion
    peak(ignoreNewlines = true): LexerToken | undefined {
```